### PR TITLE
Add retries to rename and removing directories

### PR DESCRIPTION
**Pull request by rakeshvanga**

### CODE REVIEW

File: src/utility/FileSystem.ts [hunk 1]
```diff
@@ -4,4 +4,4 @@
 import { promisify } from "util";
 import * as rimraf from "rimraf";
 
-export const rimrafAsync = promisify(rimraf);
\ No newline at end of file
+export const rimrafAsync = promisify(rimraf.sync);
```
Comment (daniv-msft) [inline src/utility/FileSystem.ts:7]: Could you please provide context for this change? Why do we need to use the sync version of rimraf here?

File: src/utility/RetryUtility.ts [hunk 2]
```diff
@@ -0,0 +1,21 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+export class RetryUtility {
+    public static async exponentialRetryAsync<T>(requestFn: () => Promise<T>, retries: number, delayInMs: number): Promise<T> {
```
Comment (lolodi) [inline src/utility/RetryUtility.ts:5]: Name this `initialDelayInMs` to be explicit on what this delay is referring to?

File: src/utility/RetryUtility.ts [hunk 3]
```diff
@@ -0,0 +1,21 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+export class RetryUtility {
```
Comment (daniv-msft) [inline src/utility/RetryUtility.ts:4]: This looks quite similar to the logic in `HttpRequestHandler`. Do you have any plans to reconcile these?

File: src/FileDownloader.ts [hunk 4]
```diff
@@ -107,15 +113,20 @@ export default class FileDownloader implements IFileDownloader {
...
         }
 
         // If the file/folder already exists, remove it now
-        await rimrafAsync(fileDownloadPath);
-        // Move the temp file/folder to its permanent location and return it
-        await fs.promises.rename(tempFileDownloadPath, fileDownloadPath);
-        return Uri.file(fileDownloadPath);
+        await rimrafAsync(fileDownloadPath, rimrafOptions);
+
+        const renameDonwloadedFileAsyncFn = async (): Promise<Uri> => {
```
Comment (amsoedal) [inline src/FileDownloader.ts:123]: nit: spelling ("Downloaded")

File: src/FileDownloader.ts [hunk 5]
```diff
@@ -107,15 +113,20 @@ export default class FileDownloader implements IFileDownloader {
...
+            // Move the temp file/folder to its permanent location and return it
+            await fs.promises.rename(tempFileDownloadPath, fileDownloadPath);
+            return Uri.file(fileDownloadPath);
+        };
+
+        return RetryUtility.exponentialRetryAsync(renameDonwloadedFileAsyncFn, retries, retryDelayInMs);
```
Comment (amsoedal) [inline src/FileDownloader.ts:129]: nice! Excited to hopefully see our reliability improve!

File: package.json [hunk 6]
```diff
@@ -23,15 +23,15 @@
     "scripts": {
         "vscode:prepublish": "webpack --mode production",
         "compile": "npm run webpack",
-        "lint": "eslint src --ext ts",
+        "lint": "eslint src --ext ts --fix",
```
Comment (rakeshvanga) [inline package.json:26]: added fix here so that no need to run it as seperate command.

**Review by daniv-msft — APPROVED** [review]
Looks good to me!

**Pull request merged**